march_rqt_gait_generator/src/march_rqt_gait_generator/gait_generator.py

```diff
@@ -225,104 +222,66 @@ def create_joint_setting(self, joint):
 
         show_velocity_markers = self.velocity_markers_check_box.isChecked()
         joint_setting_plot = JointSettingPlot(joint, self.gait.duration, show_velocity_markers)
+        joint_setting.Plot.addItem(joint_setting_plot)
 
-        self.undo_button.clicked.connect(
-            lambda: [self.undo,
-                     UserInterfaceController.update_ui_elements(
-                         joint, table=joint_setting.Table, plot=joint_setting_plot, duration=self.gait.duration,
-                         show_velocity_markers=self.show_velocity_markers_checked()),
-                     self.publish_preview()
-                     ]
-        )
+        joint_setting.Table = UserInterfaceController.update_table(
+            joint_setting.Table, joint, self.gait.duration)
+        # Disable scrolling horizontally
+        joint_setting.Table.horizontalScrollBar().setDisabled(True)
+        joint_setting.Table.horizontalHeader().setSectionResizeMode(QHeaderView.Stretch)
 
-        self.redo_button.clicked.connect(
-            lambda: [self.redo,
-                     UserInterfaceController.update_ui_elements(
-                         joint, table=joint_setting.Table, plot=joint_setting_plot, duration=self.gait.duration,
-                         show_velocity_markers=self.show_velocity_markers_checked()),
-                     self.publish_preview()
-                     ]
-        )
+        def update_joint_ui():
+            UserInterfaceController.update_ui_elements(
+                joint, table=joint_setting.Table, plot=joint_setting_plot, duration=self.gait.duration,
+                show_velocity_markers=self.velocity_markers_check_box.isChecked()),
+            self.publish_preview()
+
+        def add_setpoint(joint, time, position, button):
+            if button == QtCore.Qt.ControlModifier:
+                joint.add_interpolated_setpoint(time)
+            else:
+                joint.add_setpoint(Setpoint(time, position, 0))
+
+        self.undo_button.clicked.connect(update_joint_ui)
+        self.redo_button.clicked.connect(update_joint_ui)
 
         self.velocity_markers_check_box.stateChanged.connect(
             lambda: [joint.set_setpoints(UserInterfaceController.plot_to_setpoints(joint_setting_plot)),
-                     UserInterfaceController.update_ui_elements(
-                         joint, table=joint_setting.Table, plot=joint_setting_plot, duration=self.gait.duration,
-                         show_velocity_markers=self.show_velocity_markers_checked()),
-                     self.publish_preview()
-                     ]
-        )
+                     update_joint_ui()
+                     ])
```

Comment:
You can create a `def` for the combination of these methods like so:
```python
def l():
    joint.set_setpoints(UserInterfaceController.plot_to_setpoints(joint_setting_plot))
    update_joint_ui()
```
And then use this to connect:
```python
self.velocity_markers_check_box.stateChanged.connect(l)
joint_setting_plot.plot_item.sigPlotChanged.connect(l)
```
You might have to think of a better name :+1:

Reply:
I like the consistency of using update_joint_ui() in every connect, so I am going to keep it the way it is. Fixed the indentation though.